lib/frrevent.h

```diff
@@ -42,6 +55,92 @@ struct rusage_t {
 PREDECL_LIST(event_list);
 PREDECL_HEAP(event_timer_list);
 
+#if EPOLL_ENABLED
+struct frr_epoll_event {
+	/* the index of this frr_epoll_event instance in .event_ptrs or
+	 * .new_event_ptrs
+	 */
+	long index; // index should be -1 for a invalid frr_epoll_event
+
+	struct epoll_event ev;
+};
+
+struct fd_handler {
+	/* number of pfd that fit in the allocated space of pfds. This is a
+	 * constant and is the same for both pfds and copy.
+	 */
+	nfds_t event_size;
+
+	/* the epoll fd set to monitor for i/o. This should be coordinated with
+	 * the "events" field below
+	 */
+	int epoll_fd;
+
+	/* Below we define two arrays, namely .events and .new_events.
+	 * The function of two arrays are analogous to .pfds and .copy in old
+	 * version fd_handler which uses poll().
+	 *
+	 * .events hold existing monitored events that have been registered in
+	 * kernel by epoll_ctl() calls. Similar to .copy array in old version,
+	 * .events is updated just before every time fd_poll() is called.
+	 *
+	 * .new_events hold event update requests during the interval from last
+	 * fd_poll() call to next fd_poll() call. .new_events functions as a
+	 * holder of new event request, which is similar to .pfds array in old
+	 * version
+	 *
+	 * Both array are indexed by fd for fast update. However, during
+	 * cancel_arg_helper when canceling a thread, all frr_epoll_events
+	 * are supposed to be traversed, which is inefficient. We define two
+	 * compact arrays .event_ptrs and .new_event_ptrs for fast canceling.
+	 * Only (new_)event_count number of elements in .(new_)event_ptr are
+	 * valid, and each element refers a valid frr_epoll_event.
+	 *
+	 * frr_epoll_event arrays and ptr arrays are mutually referenced. An
+	 * frr_epoll_event reference its corresponding ptr element by its
+	 * .index field. An ptr reference its corresponding frr_epoll_event
+	 * natually since it is a pointer.
+	 */
+
+
+	/* the buffer which stores monitored fds and corresponding events,
+	 * indexed by fd
+	 */
+	struct frr_epoll_event *events;
+	/* a compact array of pointers to frr_epoll_event in .events array,
```

Comment:
so ... this approach leads to pre-allocating the absolute maximum number of epoll objects, times four. I'm concerned about that a bit - we've had issues with the memory footprint of large fd limits. Could these objects be ... linked in a list instead, for iteration? we could pre-allocate some number in advance, but then we'd save memory when not using the max number of fds.

Reply:
The reason i maintained many arrays is that i tried to be consistent with code style before, which defined a pfds/copy arrays. pfds and copy arrays are necessary, since they are arguments in poll() callings.

Indeed, when using epoll, it is unnecessary to maintain current monitoring fds and events in arrays (epoll_wait doesn't need an input array). Currently, I am trying to turn arrays u mentioned into a more compact hash table, which should be easier to understand and has less memory footprint.